### ISSUE TITLE
Success Criterion 1.4.3 - Contrast (Minimum) - Level AA

### ISSUE THREAD

**Issue by JJdeGroot** [w3c/matf#28]
Placeholder for meeting of July 17, 2024.

WCAG2ICT guidance: https://w3c.github.io/wcag2ict/#contrast-minimum

Share your thoughts for applying to mobile apps as a comment below.

**Comment by illai** [comment]
This success criterion is allegedly clear and straightforward. However, it poses a challenge when referring to the measurement units. Unlike the web on mobile, there is one platform that uses PT (points) as measurement units for text (iOS), while the other platform (Android) uses SP (Scalable Pixel) for this. Sticking to PT (points) as measurement units (which makes sense) will probably require, in the context of native mobile apps, at least a clarification and an indication of how to convert units correctly to the relevant measurement units (which I think is also missing in WCAG today)

**Comment by qbalsdon** [comment]
There might be a need to consider research on colour contrasts for smaller screens, but until that is done I am happy with this criterion remaining as is for mobile.

**Comment by qbalsdon** [comment]
@illai I'm not sure if I understand - are there units in colour contrast, or is this comment for [Touch Target Sizes](https://github.com/w3c/matf/issues/10)?

**Comment by qbalsdon** [comment]
Can we please add a condition onto this one to reference "Visual information required to identify [user interface components](https://www.w3.org/TR/WCAG22/#dfn-user-interface-components) and [states](https://www.w3.org/TR/WCAG22/#dfn-states), except for inactive components or where the appearance of the component is determined by the user agent and not modified by the author;" (sometimes called "disabled" or "not enabled") - in [1.4.11 Non text contrast](https://www.w3.org/TR/WCAG22/#non-text-contrast)

It's understandable and predictable that creators will prevent interaction by this mechanism, and we should add some info:
1. It's fine as long as state is conveyed
2. inactive components do not need to meet colour contrast rules
and best practice is to inform users why the component is inactive, but not required

**Comment by illai** [comment]
@qbalsdon The Color contrast SC has size dependencies. Contrast of at least 4.5:1 for texts blow 18 points, Contrast of at leas 3:1 for texts above 18 points. The measurement units referred on this SC are points. While it make sense to measure typography with points it also the iOS text units, while Android uses other unuts. I think that it would be important to clarify that point.

**Comment by qbalsdon** [comment]
@illai Thank you - I missed that!

**Comment by Keanem6** [comment]
Should we be include requirements for Dark mode as well?

**Comment by AlainVagner** [comment]
@illai just one question about this. As I understood in MCAG, the threshold is at 18 points for normal text and 14 points for bold (same on Apple and Android guidelines). The points you mention here are I suppose sp for Android and pt for iOS. Is that correct?
The definition of points is quite different on the web and on native mobile platforms ([1/72 of an inch on the Web](https://www.w3.org/TR/css3-values/#absolute-lengths)), can we reliably consider that they are equivalent?

I agree with you that we should clarify this point. We can evaluate a font size with @detlevhfischer 's method 
based on thresholds in css pixels, but we need also to have clear thresholds for developers in the native units.

**Comment by illai** [comment]
@AlainVagner I am basically suggesting (and this is the idea I tried to present on MGAG) to have an agreed, or an "official" furmula to convert Android and iOS units to points and vice versa, however the formula on MCAG is just framing the idea, and is very "naive" and is considering poins as a phisical value like milimeters, because this unit is coming from the print world. I do agree that there is a real difficulty in measuring it accurately due to differences in pixel densities and screen sizes of different devices.

**Comment by dotjay** [comment]
Note that WCAG 2 refers to print "points" (1pt = 1/72 of an inch) rather than iOS "points" (arbitrary levelling unit for accommodating different pixel densities, and is around 1/160 of an inch and varies slightly between device screens).

There’s a high likelihood of readers getting confused about these so I feel we need to be clear on the thresholds on this SC.

**Comment by JJdeGroot** [comment]
Discussed in today’s meeting.

<details>
<summary>11 December 2024</summary>

[Source](https://www.w3.org/2024/12/11-matf-minutes.html#7897)

~~~text
Please define large text

What is the smallest acceptable text size?

EN301549 suggests that providing dark mode will be required

<Mick> My recommendation would be to add requirement around Dark mode and remove the 'Large Text' variation.

Units of measurement is also difficult on different platforms

<julianmka> +1 Jamie on importance of font weight

<Aash> +1 to Jamie

<Aash> Thin typefaces are a problem

<julianmka> https://www.w3.org/WAI/WCAG22/Understanding/contrast-minimum#intent:~:text=Due%20to%20anti,this%20Success%20Criterion.

Jamie Some of the understanding docs discuss the thinness of font... This has more of an affect on smaller screens. Phones are read at more angles than screens. We need to think about bringing attention to visualising font on a phone

<Jon_Gibbins> +1 to Jamie’s point

<Illai> +1 to Jamie

Aash I think choice of font is part of the EN310549 and user choice (11.7)

Jamie it might be helpful to recommend a min text size

<Jamie> term: anti-aliasing

<julianmka> Could we go with the minimum sizes used in HIG and Material?

<Jon_Gibbins> Points are just confusing though, because “web points” are not the same as iOS points I think??

<Zakim> Jon_Gibbins, you wanted to clarify commenting on glare-related / dynamic environment issues surrounding mobile

Jon_Gibbins not sure if the apple pts map directly to print sizes

<julianmka> https://developer.apple.com/design/human-interface-guidelines/typography#iOS-iPadOS-Dynamic-Type-sizes

julianmka referenced the HIG

In the small print

Can someone get my magnifying glass?

ACTION: Research measuring text sizes on Android and iOS, mapping them do dp/pt and comparing with CSS points / pixels

<Mick> +1 to julianmka point

Jamie does the increased contrast apply to apps that use system fonts, but also those with custom and hybrid content

Oh sorry

<julianmka> My previous comment: We need to account for the known contrast failures in default iOS styles

julianmka I know Samsung allows for users to download their own fonts for the system, which responds to settings. My understanding is that if you are using a complete font the user settings are always correctly applied

<Jon_Gibbins> Just clarifying my thought: WCAG 2 refers to print points (1pt = 1/72 of 1in) rather than iOS points (arbitrary levelling unit for accommodating different pixel densities). There’s a high likelihood of readers getting confused about these so I feel we need to be clear on the thresholds on this SC (1.4.3).

JJ custom fonts are a little hard on iOS

@Jam'

Jamie how is hybrid content affected?

Apologies to hybrid fans, but I do think hybrid should be subject to the platform on which it appears. You don't get a pass for trying to shortcut development time

ACTION: Make a list of known default Android/iOS components that fail contrast requirements

Jamie can we review the actions?

ACTION: Consider adding thinness/thickness requirements as note

ACTION: Think about effect of system settings on contrast requirements

The font sizes?

Minimum font size

ACTION: Provide guidance for dark mode / light mode (do both need to pass contrast?)
~~~
</details>

Actions:
- [Research measuring text sizes on Android and iOS, mapping them do dp/pt and comparing with CSS points / pixels](https://www.w3.org/2024/12/11-matf-minutes.html#c1e6)
- [Make a list of known default Android/iOS components that fail contrast requirements](https://www.w3.org/2024/12/11-matf-minutes.html#0fe6)
- [Consider adding thinness/thickness requirements as note](https://www.w3.org/2024/12/11-matf-minutes.html#a2f3)
- [Think about effect of system settings on contrast requirements](https://www.w3.org/2024/12/11-matf-minutes.html#4779)
- [Provide guidance for dark mode / light mode (do both need to pass contrast?)](https://www.w3.org/2024/12/11-matf-minutes.html#fdb5)

**Comment by JJdeGroot** [comment]
Discussed in today’s meeting.

<details>
<summary>16 April 2025</summary>

[Source](https://www.w3.org/2025/04/16-matf-minutes.html#4020)

~~~text
1.4.3 Contrast (Minimum)

<pauljadam> placeholder text is text

<JJ> Thx rachaely - to filter issues on having assignees: https://github.com/w3c/matf/issues?q=is%3Aissue%20state%3Aopen%20%20has%3Aassignee

Tanya 2 q's 1. How shall we treat placeholder text, which can be treated as decorative - we reject placeholder text that doesn't meet contrast?

<pauljadam> yep designers don't like labels, which makes UIs more confusing

Tanya 2. When we are talking about color contrast, are we talking about text with no accessibility features on? We say that user settings should be respected

what if the platform does adaptive scaling - can we really fail dev's for not getting 200%? Should this be a note?

JJ this isn't about text resizing

rachaely do high contrast settings apply here?

<julianmka> System settings as sufficient techniques? [w3c/matf#67](https://github.com/w3c/matf/issues/67)

@JJ this does seem to be a separate issue

Joe_Humbert As per dynamic scaling, is there a way that testers can determine why something is not scaling? Like which unit is being used

Detlev to Tanya - We normally ignore placeholder contrast when the placeholder is redundant because there is a label. But if there is important information in the placeholer, then it needs to meet contrast. There is also a need for graphics contrast that can affect this scenario

pauljadam for placeholder, wcag does not specify "placeholder" - text is text. For contrast, some web sites allow users to increase contrast, but I think default contrast should meet WCAG.

JJ leaving the app can also be a burden on folks

pauljadam you could rely on reduce motion, but without a pause button people will think you are inaccessible

<quintinb> +1 pauljadam

respecting system settings should be below minimum

That highlight contrast...

<pauljadam> I forgot to say that the Xcode Accessibility Inspector can test for resizable text

ACTION: Define how to calculcate text size in points for Mobile Applications

pauljadam isn't that only certain versions, like 17+?

<pauljadam> don't think so

ah ty

ACTION: Determine whether high contrast mode is an acceptable way to pass 1.4.3

<pauljadam> You can't inspect the text size only you can run the Audit function
~~~
</details>

- ACTION: Define how to calculate text size in points for Mobile Applications
- ACTION: Determine whether high contrast mode is an acceptable way to pass 1.4.3

**Comment by JJdeGroot** [comment]
Related: https://github.com/w3c/wcag/issues/4290#issuecomment-2761758769

**Comment by detlevhfischer** [comment]
A [commenter on a WCAG 2.X 1.4.3 issue](https://github.com/w3c/wcag/issues/4290#issuecomment-2753302416) mentioned that 
> When we were testing mobile phone apps, the manufacturer informed us that it is very difficult to design a mobile phone to meet Success Criterion (SC) 1.4.3 without turning on assistive technologies such as high contrast mode.

When we follow up action [Make a list of known default Android/iOS components that fail contrast requirements](https://www.w3.org/2024/12/11-matf-minutes.html#0fe6) it would be nice to know whether it is difficult or trivial to change insufficient contrast values. Would this be harder in hybrid development environments?

**Comment by Phil0menaA** [comment]
Hello again,

I would also like to share some research I did on the color contrast calculation.

The current calculation method does not fully reflect human visual perception as it does not take into account:
1) The contrast polarity: Currently contrast is calculated symmetrically, but light text on dark backgrounds appears thinner due to glow, affecting readability. The [APCA model](https://apcacontrast.com/) for calculation seems to offer an interesting solution to this problem.
2) Visual vibration: Certain color combinations cause visual fatigue and cognitive load (as per [this study](https://www.sciencedirect.com/science/article/abs/pii/S0141938224000313)), which is not accounted for.

Some examples of the issue:

![Image](https://github.com/user-attachments/assets/f62d69a0-48ed-4b02-ab64-b668876b4376)

![Image](https://github.com/user-attachments/assets/e3361abe-b0a9-4726-b024-38a01e655cac)


Others have written about this too:
https://khan-tw.medium.com/wcag2-are-you-still-using-it-ui-contrast-visibility-standard-readability-contrast-f34eb73e89ee
https://typefully.com/DanHollick/wcag-3-and-apca-sle13GMW2Brp
https://uxdesign.cc/dont-rely-only-on-tools-to-pick-accessible-colour-combinations-48c12ff4a192

I think WCAG should be the opportunity to fix this too.
